### PR TITLE
fix: Support WooPay Direct Checkout in iAPI mini-cart (WC 10.4+)

### LINKED ISSUE #11431: WooPay Direct Checkout does not work in iAPI mini-cart (WC 10.4+)
## Describe the bug

The WooPay Direct Checkout flow does not intercept the "Go to checkout" button click in the mini-cart block after WooCommerce migrated it from React to the WordPress Interactivity API (iAPI). This means users are sent to the normal checkout page instead of being redirected to WooPay.

The iAPI mini-cart became the default in **WooCommerce 10.4** (Dec 10, 2025), so this affects all stores on WC 10.4+.

Related: [WOOPMNT-4079](https://linear.app/a8c/issue/WOOPMNT-4079/hook-cart-add-item-is-not-called-in-blocks-cart-in-woopay-direct)

## Root Cause

Two compounding issues in `client/checkout/woopay/direct-checkout/index.js`:

1. **Widget detection selector is missing:** `maybeObserveMiniCart()` checks for `[data-block-name="woocommerce/mini-cart"]`, but the iAPI version uses `data-wp-interactive="woocommerce/mini-cart"` instead. The function exits early.
2. **MutationObserver is incompatible with iAPI:** The observer watches for `.wc-block-components-drawer__screen-overlay` being added to the DOM. In iAPI, the overlay is server-side rendered and always in the DOM — visibility toggles via CSS classes, not DOM insertion. The observer never fires.

## To Reproduce

1. Use WooCommerce 10.4+ (iAPI mini-cart is default).
2. Enable WooPay Direct Checkout.
3. Add an item to the cart.
4. Open the mini-cart drawer.
5. Click "Go to checkout".
6. **Expected:** Redirect to WooPay checkout.
7. **Actual:** Navigates to the normal WooCommerce checkout page.

## Environment

* WooCommerce 10.4+ (iAPI mini-cart enabled by default)
* Any WooPayments version with Direct Checkout support

### PR DESCRIPTION
**Pull request by pierorocca**
Fixes #11431

#### Changes proposed in this Pull Request

The WooPay Direct Checkout flow doesn't work in the mini-cart block after WooCommerce migrated it from React to the Interactivity API (iAPI), which became the default in WC 10.4 (Dec 2025).

Business Impacts:
1. Conversion - WooPay signed-in users are getting the guest checkout experience
2. Activation - New users that enrolled into WooPay and were cookied when they verified their email may not get into the flow that turns them into an activated user.

**Root cause — two compounding issues:**

1. **Widget detection fails:** `maybeObserveMiniCart()` checks for `[data-block-name="woocommerce/mini-cart"]`, but the iAPI version renders `data-wp-interactive="woocommerce/mini-cart"` instead. The function exits early.

4. **MutationObserver is incompatible:** The observer watches for `.wc-block-components-drawer__screen-overlay` being added to the DOM. In iAPI, the overlay is server-side rendered and always present — visibility toggles via CSS classes, not DOM insertion. The observer never fires.

**Fix:** Detect the iAPI mini-cart via `[data-wp-interactive="woocommerce/mini-cart"]` and attach click listeners directly to the SSR'd checkout buttons at page load. The legacy React mini-cart path (MutationObserver) is preserved for older WC versions.

#### Testing instructions

1. Use WooCommerce 10.4+ (iAPI mini-cart is default).
2. Enable WooPay and WooPay Direct Checkout.
5. Add a product to the cart.
6. Open the mini-cart drawer (click the cart icon in the header).
7. Click "Go to checkout".
8. **Expected:** User is redirected to WooPay checkout (or WooPay login if not logged in).
9. **Before fix:** User navigates to the normal WooCommerce checkout page.

**Backward compatibility:** Stores on older WC versions (or those that revert to React mini-cart via the `woocommerce_admin_get_feature_config` filter) continue to use the existing MutationObserver path, which is unchanged.

-------------------

- [x] Run `npm run changelog` to add a changelog file, choose `patch` to leave it empty if the change is not significant.
- [x] Covered with tests (or have a good reason not to test in description ☝️)
- [x] Tested on mobile (or does not apply)

**Post merge**

- [ ] Link to testing instructions from [release testing doc](https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions) following [these instructions](https://github.com/Automattic/woocommerce-payments/wiki/How-to-write-good-manual-testing-scenarios) : _Add link here / 'QA Testing Not Applicable'_
- [ ] Add or update [critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Critical-flows) and [testing instructions for critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Testing-instructions-for-critical-flows), if applicable.
- [ ] Add what's changed (description, screenshot, demo videos etc.) to the release announcement post, if applicable.

---

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://github.com/user-attachments/assets/a296c9eb-4f7d-44b1-8602-24fcb11de6ac

### CODE REVIEW

**Comment by pierorocca** [comment]
## Code Review Report

**Verdict: APPROVE**

Minimal, well-scoped fix that adds iAPI mini-cart detection without modifying the legacy React path. All four review dimensions pass.

### Architecture Compliance — PASS
- Entirely within frontend checkout layer. No payment flow boundaries crossed.
- Implementation matches plan: detect iAPI via `[data-wp-interactive]`, attach listeners directly, preserve legacy MutationObserver.
- Reuses existing helpers — no new abstractions.

### Security — PASS
- No SQL, XSS, or CSRF concerns — purely client-side DOM detection and event listener attachment.
- All selectors are hardcoded literals. No new data flows or network calls.

### Performance — PASS
- Net positive for WC 10.4+: eliminates MutationObserver on `document.body`.
- At most 3 `querySelector` calls on `window.load` — sub-millisecond.

### Test Coverage — FLAG (non-blocking)
- Happy path covered: iAPI element in DOM → buttons found → listeners attached → correct arguments verified.
- All 12 tests pass.
- **Suggestions:**
  1. Consider adding a test for iAPI path with third-party cookies enabled (logged-in user scenario)
  2. Consider testing the null-button edge case
  3. The `toHaveBeenCalledTimes(3)` assertion could be replaced with more specific `toHaveBeenCalledWith` checks

---

🤖 Generated with [Claude Code](https://claude.com/claude-code)

File: client/checkout/woopay/direct-checkout/index.js [hunk 1]
```diff
@@ -64,14 +64,27 @@ const addFooterCartEventListener = () => {
...
-	// Check if the widget is available on the page.
+	// iAPI mini-cart (WC 10.4+): overlay & buttons are SSR'd, already in the DOM.
+	if (
+		document.querySelector(
+			'[data-wp-interactive="woocommerce/mini-cart"]'
+		)
+	) {
+		addMiniCartEventListener();
+		addFooterCartEventListener();
+		return;
```
[the remark this answers is not in the document]
Comment (pierorocca) [reply client/checkout/woopay/direct-checkout/index.js]: Fixed. Added `.filter( Boolean )` in `handleWooPayDirectCheckout` to strip null/undefined elements before they reach `addRedirectToWooPayEventListener`. This protects all callers (iAPI path, legacy path, classic cart path) from the `null.addEventListener` throw.

**Comment by rtio** [comment]
Hey Piero — nice fix on the iAPI detection. The approach of checking `[data-wp-interactive]` and attaching directly (no MutationObserver) makes total sense for the SSR'd buttons.

I ran into an issue while testing though: **the mini-cart "Go to checkout" button doesn't redirect to WooPay on the shop page** (or any non-cart page where the mini-cart is the only checkout entry point). The cart page "Proceed to checkout" button works fine.

This is the JN store I tested: 3ac4b-pb/

### Possible root cause: race condition on `isUserLoggedIn()`

The iAPI path in `maybeObserveMiniCart()` calls `addMiniCartEventListener()` and `addFooterCartEventListener()` synchronously. Each one fires `handleWooPayDirectCheckout()` (async, not awaited). Then the main load handler also fires `handleWooPayDirectCheckout()` for the regular checkout buttons.

This creates **3 concurrent `handleWooPayDirectCheckout` calls**, all hitting `await isUserLoggedIn()` simultaneously. Since `isUserLoggedIn()` uses postMessage with a shared callback on the `UserConnect` instance, concurrent calls overwrite each other's callbacks. The responses get mixed up:

| Call | Buttons | `isUserLoggedIn` result | Outcome |
|------|---------|------------------------|---------|
| Main checkout | 0 elements (shop page) | `true` ✓ | Listeners attached to empty array |
| Mini-cart button | 1 element | `false` ✗ | **No listeners attached** |
| Footer button | 1 element | `false` ✗ | **No listeners attached** |

The `true` response is consumed by the wrong caller (the empty array from the main checkout path), while the mini-cart handlers get `false` and bail without attaching listeners.

### Suggested fix

Instead of firing separate `handleWooPayDirectCheckout` calls from `maybeObserveMiniCart`, return the iAPI buttons and merge them into the main checkout buttons array. This way there's a single `handleWooPayDirectCheckout` call with all buttons:

```js
// In maybeObserveMiniCart — return iAPI buttons instead of attaching separately:
if (document.querySelector('[data-wp-interactive="woocommerce/mini-cart"]')) {
    const miniBtn = WooPayDirectCheckout.getMiniCartProceedToCheckoutButton();
    const footerBtn = WooPayDirectCheckout.getFooterMiniCartProceedToCheckoutButton();
    return [miniBtn, footerBtn].filter(Boolean);
}

// In the load handler — merge and make a single call:
const iapiMiniCartButtons = maybeObserveMiniCart() || [];
const checkoutButtons = WooPayDirectCheckout.getCheckoutButtonElements();
handleWooPayDirectCheckout([...checkoutButtons, ...iapiMiniCartButtons]);
```

I tested this locally and confirmed it resolves the race condition — all buttons get listeners in a single pass. Couldn't fully verify the WooPay redirect from localhost (iframe communication fails locally), but the listener attachment is correct.

The legacy React MutationObserver path is unaffected since it only fires one `handleWooPayDirectCheckout` at a time (on drawer open).

**Review by rtio — CHANGES_REQUESTED** [review]
Posted my findings [here](https://github.com/Automattic/woocommerce-payments/pull/11432#issuecomment-4121143296)

**Comment by pierorocca** [comment]
Fixed in e19b5cc. `maybeObserveMiniCart()` now returns the iAPI buttons instead of firing separate `handleWooPayDirectCheckout` calls. The load handler merges all buttons into a single call, eliminating the concurrent `isUserLoggedIn()` postMessage race. Legacy MutationObserver path is unaffected (returns `[]`).

**Review by rtio — APPROVED** [review]
Tested in https://ox-of-keyboards.jurassic.ninja/, this is working now.
